Sources/SourceKitLSP/SourceKitLSPServer.swift

```diff
@@ -958,7 +958,26 @@ extension SourceKitLSPServer {
   }
 
   func initialize(_ req: InitializeRequest) async throws -> InitializeResult {
-    capabilityRegistry = CapabilityRegistry(clientCapabilities: req.capabilities)
+    // If the client can handle `PeekDocumentsRequest`, they can enable the
+    // experimental client capability `"peekDocuments"` through the `req.capabilities.experimental`.
+    //
+    // The below allows the client to pass the `"peekDocuments"` flag in the
+    // `req.initializationOptions` if for some reason the client cannot pass it
+    // directly through `req.capabilities.experimental` .
+    var clientCapabilities = req.capabilities
+    if case .dictionary(let initializationOptions) = req.initializationOptions,
+      let peekDocuments = initializationOptions["peekDocuments"]
+    {
```

Comment:
Do we even need to go through `initializationOptions` here? Shouldn’t clients just declare the capabilities in `ClientCapabilities.experimental`?

Also, should we use the request name as the key here? Ie. `workspace/peekDocuments` or `sourcekit-lsp/peekDocuments`.

Reply:
I was not able to find a way to send client capabilities from the vscode-swift extension. I asked @adam-fowler to let me know how to let sourcekit-lsp know that we support "peekDocuments" and he said the following:

> If I remember right In LanguageClientManager.createLSPClient in the clientOptions you can set initialization options which are passed to the server.
> [11:56](https://swift-open-source.slack.com/archives/D072BNXPFMW/p1719599209024749)
> I discuss something similar here https://github.com/swiftlang/sourcekit-lsp/issues/797#issuecomment-1974750452

The current implementation allows the client to set "peekDocuments" through the client capabilities, and also allows an alternative way to set the same through `initializationOptions`.

@adam-fowler Let me know if I am missing something?

Or perhaps, we can eliminate the capabilities route and stick with just `initializationOptions`, but it doesn't stick with me design wise.

Let me have your thoughts on this @ahoppen @adam-fowler

Reply:
Ah, I see. I’m fine with keeping the initialization options as a secondary way to enable `peekDocument`, I would just add a comment that explains that this is a workaround for VS Code, which can’t set client capabilities from an extension for future reference.